### PR TITLE
ftb implementation + modloader refactor from array to obj

### PR DESCRIPTION
**Pull request by Ladvace**

### CODE REVIEW

File: src/common/modals/AddInstance/Import.js [hunk 1]
```diff
@@ -121,17 +121,17 @@ const Import = ({
       return;
     }
 
-    const modloader = [];
-    if (isForge) modloader.push(FORGE);
-    else if (isFabric) modloader.push(FABRIC);
-    else if (isVanilla) modloader.push(VANILLA);
+    const laoder = {};
```
Comment (kccricket) [inline src/common/modals/AddInstance/Import.js]: Spelling typo.

Comment (kccricket) [reply src/common/modals/AddInstance/Import.js]: Actually, there's a lot of "laoder" in this PR.

Comment (Ladvace) [reply src/common/modals/AddInstance/Import.js]: thanks for telling me that, I will fix that

File: src/app/desktop/utils/downloader.js [hunk 2]
```diff
@@ -11,7 +11,8 @@ const fs = fss.promises;
 export const downloadInstanceFiles = async (
   arr,
   updatePercentage,
-  threads = 4
+  threads = 4,
+  InstancePath
```
Comment (blarfoon) [inline src/app/desktop/utils/downloader.js]: This is a no-no. The complete path should be built before passing the array to this function, not after...

File: src/app/desktop/utils/downloader.js [hunk 3]
```diff
@@ -28,9 +29,15 @@ export const downloadInstanceFiles = async (
         if (counter !== 1) {
           await new Promise(resolve => setTimeout(resolve, 5000));
         }
+
+        // for FTB you need to "build" the filPath, InstancePath is passed only for FTB packs
```
Comment (blarfoon) [inline src/app/desktop/utils/downloader.js]: Again, no-no. Do this before passing the array to this function

File: src/app/desktop/utils/getInstances.js [hunk 4]
```diff
@@ -10,10 +10,35 @@ const getInstances = async instancesPath => {
         path.join(instancesPath, instance, 'config.json')
       );
       const config = await fse.readJSON(configPath);
-      if (!config.modloader) {
+      if (!config.loader) {
         throw new Error(`Config for ${instance} could not be parsed`);
       }
 
+      if (Array.isArray(config.modloader)) {
+        const loaderType = config.modloader[0];
```
Comment (blarfoon) [inline src/app/desktop/utils/getInstances.js]: Use array destructuring please..

File: src/app/desktop/utils/getInstances.js [hunk 5]
```diff
@@ -10,10 +10,35 @@ const getInstances = async instancesPath => {
         path.join(instancesPath, instance, 'config.json')
       );
       const config = await fse.readJSON(configPath);
-      if (!config.modloader) {
+      if (!config.loader) {
         throw new Error(`Config for ${instance} could not be parsed`);
       }
 
+      if (Array.isArray(config.modloader)) {
```
Comment (blarfoon) [inline src/app/desktop/utils/getInstances.js:15]: Add a comment here to describe what's going on

File: src/common/modals/AddInstance/Import.js [hunk 6]
```diff
@@ -121,17 +121,17 @@ const Import = ({
...
 
-    const modloader = [];
-    if (isForge) modloader.push(FORGE);
-    else if (isFabric) modloader.push(FABRIC);
-    else if (isVanilla) modloader.push(VANILLA);
+    const loader = {};
+
+    if (isForge) loader.loaderType = FORGE;
+    else if (isFabric) loader.loaderType = FABRIC;
+    else if (isVanilla) loader.loaderType = VANILLA;
```
Comment (blarfoon) [inline src/common/modals/AddInstance/Import.js]: vanilla needs to be the default value here

File: src/common/modals/AddInstance/InstanceName.js [hunk 7]
```diff
@@ -126,77 +127,102 @@ const InstanceName = ({
...
-          manifest.minecraft.version,
-          version[1],
-          version[2]
-        ];
+      } else if (version?.loaderType === VANILLA) {
+        const loader = {
+          loaderType: version?.loaderType,
+          mcVersion: manifest.minecraft.version,
+          loaderVersion: version?.loaderVersion,
+          fileId: version?.fileId
```
Comment (blarfoon) [inline src/common/modals/AddInstance/InstanceName.js:174]: why fileId for vanilla?

File: src/common/modals/AddInstance/InstanceName.js [hunk 8]
```diff
@@ -206,37 +232,59 @@ const InstanceName = ({
...
-        dispatch(addToQueue(localInstanceName, modloader, manifest));
+        };
+
+        dispatch(addToQueue(localInstanceName, loader, manifest));
+      } else if (version?.loaderType === FABRIC) {
+        const loader = {
+          loaderType: version?.loaderType,
+          mcVersion: manifest.minecraft.version,
+          loaderVersion: manifest.minecraft.modLoaders[0].yarn,
+          fileId: manifest.minecraft.modLoaders[0].loader
```
Comment (blarfoon) [inline src/common/modals/AddInstance/InstanceName.js:252]: I am confused, why fileId equal to the fabric loader?

File: src/common/modals/InstanceManager/Overview.js [hunk 9]
```diff
@@ -269,34 +270,34 @@ const Overview = ({ instanceName, background, manifest }) => {
             title="Minecraft Version"
             color={props => props.theme.palette.colors.jungleGreen}
             instanceName={instanceName}
-            defaultValue={config?.modloader}
+            defaultValue={config?.loader}
```
Comment (blarfoon) [inline src/common/modals/InstanceManager/Overview.js:300]: an object cannot be a default value...

Comment (blarfoon) [reply src/common/modals/InstanceManager/Overview.js:300]: Also, what's the point of having a default value in a Card?

File: src/common/modals/ModsBrowser.js [hunk 10]
```diff
@@ -160,147 +253,106 @@ const Cell = ({
...
+      </RowContainer>
+    );
+  });
 
   return (
     <InfiniteLoader
       isItemLoaded={isItemLoaded}
-      itemCount={itemCount}
+      itemCount={itemCount !== 0 ? itemCount : 40}
+      // loadMoreItems={() => loadMoreItems()}
```
Comment (blarfoon) [inline src/common/modals/ModsBrowser.js]: Please remove this if it's not needed

File: src/common/modals/ModsBrowser.js [hunk 11]
```diff
@@ -160,147 +253,106 @@ const Cell = ({
...
       loadMoreItems={() => loadMoreItems(searchQuery)}
-      threshold={20}
+      // threshold={20}
```
Comment (blarfoon) [inline src/common/modals/ModsBrowser.js]: Why did you remove the threshold?

File: src/common/reducers/actions.js [hunk 12]
```diff
@@ -1417,14 +1425,136 @@ export function downloadForge(instanceName) {
...
     await fse.remove(tempInstaller);
   };
 }
 
-export function processManifest(instanceName) {
+export function processFTBManifest(instanceName) {
+  return async (dispatch, getState) => {
+    const state = getState();
+    const { manifest } = _getCurrentDownloadItem(state);
+    // const concurrency = state.settings.concurrentDownloads;
```
Comment (blarfoon) [inline src/common/reducers/actions.js]: Why should this function not respect the cuncurrent downloads setting?

File: src/common/reducers/actions.js [hunk 13]
```diff
@@ -1417,14 +1425,136 @@ export function downloadForge(instanceName) {
...
+            await new Promise(resolve => setTimeout(resolve, 5000));
+          }
+          try {
+            const ext = item.name.split(/\.(?=[^.]+$)/)[1];
+            const isJarFile = ext === 'jar';
+            if (isJarFile) {
+              const filePath = path.join(instancePath, item.path, item.name);
+
+              const hash = await getFileMurmurHash2(filePath);
+              const { data } = await getAddonsByFingerprint([hash]);
```
Comment (blarfoon) [inline src/common/reducers/actions.js]: This is highly inefficient because you need to make 1 API request for each mod. You should first calculate all the hashes and then only make a bunch of requests with an array of hashes to minimize the requests.

File: src/common/reducers/actions.js [hunk 14]
```diff
@@ -1417,14 +1425,136 @@ export function downloadForge(instanceName) {
...
+      updateInstanceConfig(instanceName, config => {
+        return {
+          ...config,
+          mods: [...(config.mods || []), ...modManifests]
+        };
+      })
+    );
+
+    await new Promise(resolve => {
+      // Force premature unlock to let our listener catch mods from override
```
Comment (blarfoon) [inline src/common/reducers/actions.js]: There are no overrides in FTB, what's the point of this?

File: src/common/reducers/actions.js [hunk 15]
```diff
@@ -1417,14 +1425,136 @@ export function downloadForge(instanceName) {
...
+          'installing.lock'
+        ),
+        err => {
+          if (err) console.error(err);
+          resolve();
+        }
+      );
+    });
+
+    // await fse.remove(addonPathZip);
```
Comment (blarfoon) [inline src/common/reducers/actions.js]: ??

**Review by blarfoon — CHANGES_REQUESTED** [review]
I also did not understand the difference between `source` and `loaderType`, could you clarify that?

**Comment by blarfoon** [comment]
It's on the right path. Keep working on it tho 🎉

**Review by blarfoon — APPROVED** [review]
